ddd/ddd-03-fd-blocking.DISCUSSION.md

```diff
@@ -0,0 +1,33 @@
+ddd-03-fd-blocking
+==================
+
+This proposed diff demonstrates the theoretical minimal changes for the example:
+
+    - TLS_client_method is changed to QUIC_client_method
+    - SOCK_STREAM, IPPROTO_TCP is changed to SOCK_DGRAM, IPPROTO_UDP
+
+This just about works, just. It just about works and serious thought should be
+given to whether we actually want to support this, but it is possible:
+
+    - The destination UDP endpoint is set via connect(). Connection migration
+      can be handled by subsequent calls to connect() which changes the
+      destination address, or by using sendto() etc. The latter is probably a
+      better idea anyway so connect() can be seen as just a highly backwards
+      compatible way to pass the destination endpoint to libssl. Once the FD
+      is passed to libssl, libssl can obtain the destination endpoint if
+      it needs to know by calling getpeername(3).
+
+    - As for ddd-01, it is assumed that the application spends a lot of time
+      in calls to libssl (as it is a blocking application) or at least calls it
+      regularly. Since synchronous users of libssl are probably fairly simple
+      this may be true quite often. If it is not the case, either a
+      BIO_pump/SSL_pump call must be made regularly via some means, or
+      thread-assisted operation should be used.
```

Comment:
It's more problematic than this. There is another problem of being blocked on read/readmsg() call on a blocking socket - the workaround could be to always temporarily set the socket to nonblocking mode internally before we call read/readmsg() on the socket and use poll() on the socket with a timeout. Probably having an internal thread could be another workaround.

Reply:
There are a few options here. In thread-assisted mode we have a lot of options. For the non-thread-assisted case:

- We could always poll before calling recvmsg to guarantee it doesn't block on the next call.
- We could use recvmsg with MSG_DONTWAIT where it is supported.
- We could set the FD as nonblocking. This could be a fallback for platforms which can't support MSG_DONTWAIT. This may have application-visible side effects if the application passed the FD, but it seems like it would be extremely unusual for an application to give a FD to libssl but also try and do things on it which are affected by O_NONBLOCK.